### DISCUSSION TITLE
Question: Thread safety

### DISCUSSION THREAD

**Discussion by senthilkvs** [CESNET/libyang#2063] [Q&A] [answered]
Hi, 
Is it considered safe to have multiple threads, where each thread possesses its own context, schema, and attached data ?
Is it possible for the owning threads to validate and edit data nodes under each of their respective contexts in parallel?

Thanks you !

**Comment by jktjkt** [comment, marked as answer]
[yes](https://netopeer.liberouter.org/doc/libyang/devel/html/howto_threads.html)

**Comment by senthilkvs** [comment]
Thank you for the response.  Have one more question. 
Can it be safely assumed that functions along the lines of editing the data and validating the data are written as reentrant ?

**Reply by jktjkt** [reply]
Since even `malloc()` is not reentrant, and tree manipulation might allocate memory, nope, these are not reentrant. What problem are you trying to solve?

**Comment by senthilkvs** [comment]
Thank you for the response. 

**Below is my use-case:** 
I am developing a system that manages yang data for DUTs (Devices under Test) across multiple virtual spaces. Each virtual space is responsible for a specific set of DUTs. 
The system receives yang data from the admin, which may contain configurations for multiple virtual spaces simultaneously. It is essential for the system to validate these incoming data and store them in the appropriate virtual spaces.

I am considering implementing a separate libyang context for each virtual space, even though the modules will remain the same for all. This approach would allow for a distinct data tree for each virtual space. When processing configurations for multiple virtual spaces, my intention is to handle validation and storage in parallel using threads. This strategy would ensure computational efficiency throughout the entire system.

**Reply by michalvasko** [reply]
If you have separate data trees all parsed in a single context, they can all be validated in parallel. The context is only being read.

**Reply by senthilkvs** [reply]
Are you suggesting to go with single data context ?  I would like to perform both writing changes to the data tree and validating across multiple virtual spaces in parallel.  From that perspective, would it make sense to use context per virtual-space ?
Furthermore, since the similar set of schema will be used in all virtual spaces, maintaining a single context could lead to collisions between data instances in the tree for the same schema.

**Reply by michalvasko** [reply]
Like I said, as long as you have **completely separate** data trees (`struct lyd_node *`), you can validate them all **in parallel** even if they are using a single context (`struct ly_ctx *`) as long as you are not going to modify it anymore. Data validation only reads from the schema nodes/context so it is safe.

**Reply by jktjkt** [reply]
the "libyang context" is something which manages the schema. If you have the exact same schema across multiple data instances/pools/heaps/whatever (which you [said](https://github.com/CESNET/libyang/discussions/2063#discussioncomment-6493588) you do), *and* if you do not change the schemas in any way once you've started working with data from that context, then it's OK to reuse the same context. One context supports independent "data forests" just fine. For example, there's no way to "ask the context for the data", etc.

All this is well explained in the documentation, so I suggest to re-read that starting with the introduction *and* threading concepts *and* all the parts that you're going to be using in your application.

**Reply by senthilkvs** [reply]
Thanks much for the response. I got good amount of clarity. Will go through the documents again as well to fine tune my understanding,
